packages/server/src/lockers/GCSLockFile.ts

```diff
@@ -0,0 +1,129 @@
+import {Bucket, File, FileMetadata} from '@google-cloud/storage'
+
+type MetaGeneration = string | number | undefined
+
+export type GCSLockFileMetadata = FileMetadata & {
+  /**
+   * The lock file expires at this time (in ms) if its not refreshed.
+   */
+  exp: number
+}
+
+/**
+ * Handles communication with GCS.
+ */
+class GCSLockFile {
+  /**
+   * Name of the file in the bucket.
+   */
+  protected name: string
+  /**
+   * GCS File instance for the lock.
+   */
+  protected lockFile: File
+  /**
+   * GCS File instance for release request.
+   */
+  protected releaseFile: File
+  /**
+   * The last known metageneration of the file. If it does not match the GCS metageneration, this lockfile has been deleted and another instance has already created a new one.
+   */
+  protected currentMetaGeneration: MetaGeneration
+
+  constructor(bucket: Bucket, name: string) {
+    this.name = name
+    this.lockFile = bucket.file(name)
+    this.releaseFile = bucket.file(name + '.release')
+  }
+  /**
+   * Create the lockfile with the specified exp time. Throws if the file already exists
+   */
+  public async create(exp: number) {
+    const metadata: GCSLockFileMetadata = {
+      exp,
+      cacheControl: 'no-store',
+    }
+
+    await this.lockFile.save('', {
+      preconditionOpts: {ifGenerationMatch: 0},
+      metadata,
+    })
+    this.currentMetaGeneration = 0
+  }
+
+  /**
+   * Fetch metadata of the lock file.
+   */
+  public async getMeta() {
+    return (await this.lockFile.getMetadata())[0] as GCSLockFileMetadata
+  }
+
+  /**
+   * Refresh our own lockfile. Throws if it does not exist or the file is modified by another instance.
+   */
+  public async refreshOwn(exp: number) {
+    const metadata: GCSLockFileMetadata = {
+      exp,
+    }
+    const res = await this.lockFile.setMetadata(metadata, {
+      ifMetaGenerationMatch: this.currentMetaGeneration,
+    })
+    this.currentMetaGeneration = res[0].metageneration
+  }
+  /**
+   * Check if a release request has been submitted to our own lockfile. Throws if it does not exist or the file is modified by another instance.
+   */
+  public async checkOwnReleaseRequest() {
+    const meta = await this.getMeta()
+    if (meta.metageneration !== this.currentMetaGeneration) {
+      throw new Error('This lockfile has been already taken by another instance.')
+    }
+
+    const releaseRequestExists = (await this.releaseFile.exists())[0]
+    return releaseRequestExists
+  }
+
+  /**
+   * Delete our own lockfile if it still exists.
+   */
+  public async deleteOwn() {
+    try {
+      await this.deleteReleaseRequest()
+      await this.lockFile.delete({ifGenerationMatch: this.currentMetaGeneration})
```

Comment:
Would it make sense to do this in parallel?

Reply:
It makes sense to delete the release request file first. Otherwise, the lock might be acquired by a new owner while the old release request file still exists.